### DISCUSSION TITLE
Unique ID mesh instances to optimize performance large model

### DISCUSSION THREAD

**Discussion by BIMWB** [zeux/meshoptimizer#258] [Q&A]
We are using GLTFPACK to optimize a file with many meshes and draw calls, that currently leads to a too low FPS. Important to note is that we need to track the unique ID of each object (currently stored in the object name). Many objects are repeating objects, but have an unique ID and therefore are currently unique meshes. 

After optimizing the file with GLTFPACK, the performance is great but we lose the unique ID's of the meshes due to the instances that are being created and missing nodes. When using -kn, the structure remains intact but the performance gain is gone of course. 

Is there something 'in between', that sort of creates instances, but still keeps track of the object names (or ID's) for each instance based on its location? 

Or any other suggestions? 

I've the feeling that the following two discussions are about the same topic
- https://github.com/zeux/meshoptimizer/issues/165
- https://github.com/KhronosGroup/glTF/issues/1699

Below you can find some data before and after optimizing 

Before
 
![before](https://user-images.githubusercontent.com/81398924/112568973-591a7000-8e1e-11eb-9f86-b94fafa3c527.png)

After (while loosing the unique ID)
 
![after](https://user-images.githubusercontent.com/81398924/112568995-60417e00-8e1e-11eb-87fb-b4636ae08f7a.png)

**Comment by zeux** [comment]
The core issue is that there's not really a preagreed upon schema for specifying extra per-instance data.

For example, gltfpack can easily encode object names as an array of strings somewhere, but it's not clear what the format/naming of this data should be. Short of an existing extension we could do smth like extras.instanceNames to match existing practice of using extras.targetNames for morph targets I guess, although I don't know whether extras are accessible in Babylon.JS after parsing.
